common/src/sphere_collider.rs

```diff
@@ -0,0 +1,800 @@
+        let mut endpoint = RayEndpoint {
+            tanh_distance,
+            hit: None,
+        };
+        find_vertex_collision(ctx, &mut endpoint);
+        endpoint
+    }
+
+    fn test_face_collision(ctx: &ChunkShapeCastingContext, t_axis: usize, tanh_distance: f32) {
+        let endpoint = chunk_shape_cast_wrapper(ctx, tanh_distance);
+        assert_endpoints_hit_and_eq(
+            &endpoint,
+            &find_face_collision_wrapper(ctx, t_axis, tanh_distance),
+        );
+        sanity_check_normal(ctx, &endpoint);
+    }
+
+    fn test_edge_collision(ctx: &ChunkShapeCastingContext, t_axis: usize, tanh_distance: f32) {
+        let endpoint = chunk_shape_cast_wrapper(ctx, tanh_distance);
+        assert_endpoints_hit_and_eq(
+            &endpoint,
+            &find_edge_collision_wrapper(ctx, t_axis, tanh_distance),
+        );
+        sanity_check_normal(ctx, &endpoint);
+    }
+
+    fn test_vertex_collision(ctx: &ChunkShapeCastingContext, tanh_distance: f32) {
+        let endpoint = chunk_shape_cast_wrapper(ctx, tanh_distance);
+        assert_endpoints_hit_and_eq(
+            &endpoint,
+            &find_vertex_collision_wrapper(ctx, tanh_distance),
+        );
+        sanity_check_normal(ctx, &endpoint);
+    }
+
+    /// Check that the two endpoints contain a hit and are equal to each other. Useful for
+    /// ensuring that a particular intersection type is detected by the general `chunk_shape_cast`
+    /// method.
+    fn assert_endpoints_hit_and_eq(endpoint0: &RayEndpoint, endpoint1: &RayEndpoint) {
+        assert_eq!(endpoint0.tanh_distance, endpoint1.tanh_distance);
+        assert!(endpoint0.hit.is_some());
+        assert!(endpoint1.hit.is_some());
+        assert_eq!(
+            endpoint0.hit.as_ref().unwrap().normal,
+            endpoint1.hit.as_ref().unwrap().normal
+        );
+    }
+
+    /// Ensures that the normal is pointing outward, opposite the ray direction.
+    fn sanity_check_normal(ctx: &ChunkShapeCastingContext, endpoint: &RayEndpoint) {
+        // The ray we care about is after its start point has moved to the contact point.
+        let ray = math::translate(
+            &ctx.ray.position,
+            &math::lorentz_normalize(&ctx.ray.ray_point(endpoint.tanh_distance)),
+        ) * ctx.ray;
+
+        // The ray should be in the normal's coordinate system.
+        let ray = ctx.transform * &ray;
+
+        let normal = endpoint.hit.as_ref().unwrap().normal;
+
+        // Project normal to be perpendicular to the ray's position
+        let corrected_normal =
+            math::lorentz_normalize(&(normal + ray.position * math::mip(&normal, &ray.position)));
+
+        // Check that the normal and ray are pointing opposite directions
+        assert!(math::mip(&corrected_normal, &ray.direction) < 0.0);
+    }
+
+    /// Tests that a suitable collision is found when approaching a single voxel from various angles.
+    #[test]
+    fn chunk_shape_cast_examples() {
```

Comment:
A few negative cases here wouldn't go amiss.

Reply:
I do have negative cases in separate tests, but I can add one here as well.